src/journal/journald-syslog.c

```diff
@@ -328,29 +333,37 @@ void server_process_syslog_message(
         struct iovec iovec[N_IOVEC_META_FIELDS + 6];
         unsigned n = 0;
         int priority = LOG_USER | LOG_INFO;
-        _cleanup_free_ char *identifier = NULL, *pid = NULL;
+        _cleanup_free_ char *identifier = NULL, *pid = NULL, *m = NULL;
+        char *syslog_message;
         const char *orig;
 
         assert(s);
         assert(buf);
 
         orig = buf;
+
         syslog_parse_priority(&buf, &priority, true);
 
         if (s->forward_to_syslog)
-                forward_syslog_raw(s, priority, orig, ucred, tv);
+                forward_syslog_raw(s, priority, orig, buf_len, ucred, tv);
+
+        syslog_message = m = memdup(buf, buf_len);
```

Comment:
> I guess that instead of memdup I could use alloca + memcpy?

Well, @poettering suggested `strdupa`. Internally, it calls `alloca`, `memcpy`

The main adventage of https://github.com/systemd/systemd/pull/4856#issuecomment-267976998 is
```c
if (p != e) 
        buf = strndupa(buf, p - buf); /* replace buffer, if there was something to chop off */
```
i.e.
> This way, we will copy the buffer only if we have to truncate (which hopefully is the exception)